test/unit/argument_parser/format_parse_validators_test.cpp

```diff
@@ -1134,6 +1134,84 @@ TEST(validator_test, regex_validator_error)
     EXPECT_THROW(parser4.parse(), seqan3::validation_error);
 }
 
+TEST(validator_test, chaining_validators_common_type)
+{
+    // chaining integral options stay integral
+    {
+        int max_int = std::numeric_limits<int>::max();
+        std::vector v_int{1, 2, 3, max_int};
+        std::vector v_unsigned{4u, static_cast<unsigned>(max_int)};
+
+        EXPECT_TRUE((std::same_as<std::vector<int>, decltype(v_int)>));
+        EXPECT_TRUE((std::same_as<std::vector<unsigned>, decltype(v_unsigned)>));
+
+        seqan3::value_list_validator validator_int{v_int};
+        seqan3::value_list_validator validator_unsigned{v_unsigned};
+
+        EXPECT_TRUE((std::same_as<seqan3::value_list_validator<int>, decltype(validator_int)>));
+        EXPECT_TRUE((std::same_as<int, decltype(validator_int)::option_value_type>));
+
+        EXPECT_TRUE((std::same_as<seqan3::value_list_validator<unsigned>, decltype(validator_unsigned)>));
+        EXPECT_TRUE((std::same_as<unsigned, decltype(validator_unsigned)::option_value_type>));
+
+        auto validator = validator_int | validator_unsigned;
+
+        EXPECT_TRUE((std::same_as<unsigned, std::common_type_t<int, unsigned>>));
+        EXPECT_TRUE((std::same_as<unsigned, decltype(validator)::option_value_type>));
+
+        // max_int is part of both validators
+        EXPECT_NO_THROW(validator_int(max_int));
+        EXPECT_NO_THROW(validator_unsigned(max_int));
+        EXPECT_NO_THROW(validator(max_int));
+    }
+
+    // chaining mixed arithmetic options will be highest common arithmetic type
+    {
+        // note: this number is not representable by double and multiple integer values represent the same double value
+        int64_t max_int64 = std::numeric_limits<int64_t>::max();
+        EXPECT_EQ(static_cast<double>(max_int64), static_cast<double>(max_int64 - 1));
```

Comment:
I didn't know this! This is the weirdest line of code. But nice!

Reply:
It really is a weird line of code; but I hope the comment could help to understand it :)